### PR TITLE
Automatically prepare indices for splitting

### PR DESCRIPTION
**Pull request by s1monw**
Today we require users to prepare their indices for split operations.
Yet, we can do this automatically when an index is created which would
make the split feature a much more appealing option since it doesn't have
any 3rd party prerequisites anymore.

This change automatically sets the number of routinng shards such that
an index is guaranteed to be able to split once into twice as many shards.
The number of routing shards is scaled towards the default shard limit per index
such that indices with a smaller amount of shards can be split more often than
larger ones. For instance an index with 1 or 2 shards can be split 10x
(until it approaches 1024 shards) while an index created with 128 shards can only
be split 3x by a factor of 2. Please note this is just a default value and users
can still prepare their indices with `index.number_of_routing_shards` for custom
splitting.

NOTE: this change has an impact on the document distribution since we are changing
the hash space. Documents are still uniformly distributed across all shards but since
we are artificually changing the number of buckets in the consistent hashign space
document might be hashed into different shards compared to previous versions.

This is a 7.0 only change.

### CODE REVIEW

File: qa/smoke-test-reindex-with-all-modules/src/test/resources/rest-api-spec/test/update_by_query/10_script.yml
Comment (bleskes) [inline qa/smoke-test-reindex-with-all-modules/src/test/resources/rest-api-spec/test/update_by_query/10_script.yml]: why is this relevant?

**Review by bleskes — COMMENTED** [review]
I did an initial pass and left some comments. The basic setup looks great to me. I also like and see the need for ability to "reconfigure" the number of routing shards when the source shard is one (as otherwise what people chose may not be a divisor of our automatic routing shard value).  I'm a bit concerned by the complexity of trying to leave the routing shards as they were if possible and otherwise calculating a fresh one. Why can't we go with the simple option of *always* re-calculating number of routing shards (and allowing to set it) if the source has 1 shard?

File: qa/smoke-test-reindex-with-all-modules/src/test/resources/rest-api-spec/test/update_by_query/10_script.yml
Comment (s1monw) [reply qa/smoke-test-reindex-with-all-modules/src/test/resources/rest-api-spec/test/update_by_query/10_script.yml]: because the doc order changes and then we get `update` instead of `junk` and that is invalid too.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: :)

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: I think it's worth sharing the 1024 constant with the max value in `buildNumberOfShardsSetting` to be clearer where this comes from?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: It feels a bit weird to me that this will generate numbers that are greater than the maximum number of shards. Should we change the formula a bit so that the result is always in 513...1024 when the number of shards is in 1..512?

This probably deserves some comments as well, eg. I presume that the fact we do `Math.max(1, ...)` instead of `Math.max(0, ...)` is to make sure that we can always split at least once?

File: core/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java]: should we assert that intRatio is a power of two by checking that `intRatio & (intRatio - 1)` is zero? (or `intRatio == Integer.highestOneBit(intRatio)` if you find it easier to read

**Review by jpountz — CHANGES_REQUESTED** [review]
The change looks good to me overall, but I think we should add some comments to make it easier to understand. Also maybe add a note to the migration guide since this will change how documents are spread across shards when the routing factor is not set at index-creation time?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: there is a comment in the line below?!

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: I am not sure I understand what you mean

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: what woud you name it `a, b, c` :D I mean I can do step1, step2, step3 not sure how much it will by us ot's jiust a test

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: ++

**Comment by s1monw** [comment]
> I did an initial pass and left some comments. The basic setup looks great to me. I also like and see the need for ability to "reconfigure" the number of routing shards when the source shard is one (as otherwise what people chose may not be a divisor of our automatic routing shard value). I'm a bit concerned by the complexity of trying to leave the routing shards as they were if possible and otherwise calculating a fresh one. Why can't we go with the simple option of always re-calculating number of routing shards (and allowing to set it) if the source has 1 shard?

+1 I will work on a change for this

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: yeah I will do that.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: not sure how you envisioned this to work?

Comment (jpountz) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: I was just thinking of having a `public static final int DEFAULT_MAX_NUM_SHARDS = 1024` somewhere, and use it in `buildNumberOfShardsSetting` and here: `int base = Math.log(DEFAULT_MAX_NUM_SHARDS)/Math.log(2)`

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (bleskes) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: yeah, I get this means we only do the new behavior until the cluster is fully upgraded, but I don't see why we care? I mean, if the master is a 7.0.0 master, we can start creating indices with a different hashing logic and not worry about it?

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: ❤️

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (bleskes) [reply core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: The test now start with 1 shard source and the split twice. both of these always have a valid number of routing shards in the source index. I think the interesting part of the test is see how we reset the number of routing shard. For example, start with a 3 shards index. Shrink to 1 (number of routing shards stays 3) then split to say, 2. Does that help?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (s1monw) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: it's mainly for testing purposes and BWC behavior being more predictable otherwise some rest tests will randomly fail

File: core/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java]: done

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: 😍

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: missing end of comment?

File: core/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java]: s/ration/ratio/

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: I think it'd be better for results to be in 513..1024 than 512..1023 by ceiling the log? Feel free to ignore but I'd do the following:

```java
// We use as a default number of routing shards the higher number that can be expressed as {@code numShards * 2^x`} that is less than or equal to the maximum number of shards: 1024.
int log2MaxNumShards = 10; // logBase2(1024)
int log2NumShards = 32 - Integer.numberOfLeadingZeros(numShards - 1); // ceil(logBase2(numShards))
int numSplits = log2MaxNumShards - log2NumShards;
numSplits = Math.max(1, numSplits); // Ensure the index can be split at least once
return numShards * 1 << numSplits;
```

And then in tests make sure that the value is in 513..1024 for any number of shards in 0..512, and equal to numShards*2 for any number of shards that is greater than 512?

**Review by jpountz — COMMENTED** [review]
I left a comment about the computation of the default number of routing shards.

**Comment by s1monw** [comment]
@bleskes @jpountz it's ready for another round

**Review by jpountz — APPROVED** [review]
LGTM, but please wait for Boaz's review before merging as he had different concerns from me.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java
Comment (bleskes) [reply core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: oh well :)

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/admin/indices/create/SplitIndexIT.java]: +1

**Review by bleskes — APPROVED** [review]
LGTM. Thanks for this. It must have a pain to hunt down all the test failures.

**Pull request merged**
